### PR TITLE
first pass at procedures

### PR DESCRIPTION
**Pull request by brennanjl**
Threw together a first pass of procedure syntax. I think we need to make some larger changes to the docs since procedures are becoming preferred to actions, but this will work to introduce someone to the syntax.

### CODE REVIEW

File: docs/kuneiform/procedures.md [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+```ts
+procedure declare_vars() public {
+    // declare and assign new int var:
+    $var1 int := 1;
+
+    // declare a new var to be assigned to later
+    $var2 text;
+
+    // assign to $var2. type cast to match type text:
+    $var2 := $var1::text;
```
Comment (jchappelow) [inline docs/kuneiform/procedures.md:53]: A question that are reader is likely to have is, what type casts are valid?  For instance, what if someone did `$var3 text := 'nan'` and then `$var3::int`?

File: docs/kuneiform/procedures.md [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+```ts
+procedure do_math($num1 int, $num2 int) public {
+    // add and divide, with precedence:
+    $avg int := ($num1 + $num2)/2;
+
+    // modulo:
+    $mod int := $num2%10;
+
+    // negate
```
Comment (jchappelow) [inline docs/kuneiform/procedures.md:69]: Pretty obvious, but `// negate, multiply` maybe?

Comment (jchappelow) [reply docs/kuneiform/procedures.md:69]: A secondary question that comes to mind is order of operations / precedence.  e.g. what happens without the parentheses, and what are all the other precedence?

File: docs/kuneiform/procedures.md [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+}
+```
+
+### Executing SQL
+
+SQL statements can be executed at any point in a procedure. In the below example, we generated a deterministic UUID using a random UUID namespace and the transaction ID. We then use the generated id as the user's primary key.
+
+```ts
+procedure create_user($name text, $age int) public {
+    $id uuid := generate_uuid_v5('455f60aa-0569-4aaa-8469-63be2ec4dd96'::uuid, @txid)
```
Comment (jchappelow) [inline docs/kuneiform/procedures.md:80]: Should we list the recognized and allowed functions like `generate_uuid_v5`  for procedures and actions?  Or is there a page under "sql-as-understood-by-kwil" that should be updated?

File: docs/kuneiform/procedures.md [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+- `array_append(type[], type) type[]`: appends a value to the end of an array.
+- `array_prepend(type, type[]) type[]`: prepends a value to the beginning of an array.
+- `array_cat(type[], type[]) type[]`: concatenates two arrays together.
+- `array_length(type[]) int`: returns the length of the array.
+
+**Arrays are 1-indexed.**
+
+```ts
+procedure handle_array() public view returns (val text) {
+    $arr text[] := ['hello', 'world'];
```
Comment (jchappelow) [inline docs/kuneiform/procedures.md:101]: Mostly unrelated to this `handle_array` example, but do we support fixed length arrays like `int[16]` as postgres does, or would that have no value in procedures or procedure returns?

File: docs/kuneiform/procedures.md [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+}
+```
+
+#### Procedure Loops
+
+You can iterate over any SQL statement as well. The following examples reference an example table `users`, with columns `id`, `name`, and `age`.
+
+```ts
+procedure iterate_over_sql() public {
+    for $row in SELECT id, name, age from users {
```
Comment (jchappelow) [inline docs/kuneiform/procedures.md:163]: SQL formatting is a little inconsistent.  `SELECT` is caps, but `from` is not.  Would look better in caps.

In the subsequent example with `iterate_over_delete` it's all lower case (`delete from users ...`), which imo makes it harder to read when it follows `in`.

Then in `iterate_over_users` it looks proper with `for $row in SELECT * FROM get_users() {`

File: docs/kuneiform/procedures.md [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+```ts
+procedure get_users() public {
+    $count int := 0;
+    for $row in SELECT * FROM users {
+        // do something
+
+        // break if too many records have been read
+        $count := $count + 1;
+        if $count >= 50 {
+            break
```
Comment (jchappelow) [inline docs/kuneiform/procedures.md:210]: I assume `break` works like any other sensible language with nested loops by breaking only the inner loop?  Also assuming nested loops are ok?

Do we have a `continue` do programmers just code around that with else and more indents?  That's fine, just curious.

File: docs/kuneiform/procedures.md [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        }
+    }
+}
+```
+
+
+
```
Comment (jchappelow) [inline docs/kuneiform/procedures.md:217]: nit, a bunch of extra lines here.

File: docs/kuneiform/procedures.md [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+```ts
+procedure get_users() public view returns table(id uuid, name text, age int) {
+    return select * from users;
+}
+```
+
+#### Return Next
+
+If a procedure returns a table and should to selectively return values, you can use `return next` to return a record inside of a for loop. Unlike `return`, `return next` will not exit the procedure, but instead continue. In the below example, we query a table `prices` with columns `value` and `day`, and return the net change from each day. We assume that our first day is day 1. If a day is missing, it returns an error.
```
Comment (jchappelow) [inline docs/kuneiform/procedures.md:257]: I'm having trouble parsing "and should to selectively return values" in this context.  Is "to" just not supposed to be there?

Comment (jchappelow) [reply docs/kuneiform/procedures.md:257]: "a for loop" => "a `for` loop"

File: docs/kuneiform/procedures.md [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+        if $row.day != $last_day + 1 {
+            error(format('missing day %d', $last_day+1));
+        }
+
+        // add the change to the result set
+        return next $row.day, $row.value - $last_value;
+        // update our last seen value
+        $last_value := $row.value;
+    }
+}
```
Comment (jchappelow) [inline docs/kuneiform/procedures.md:274]: I guess any procedure with a `return next` somewhere has some implicit "`return collected`" at the end.  I'm just noting that there's no `return` at the end, but the `for` could never call a `return next`, and the procedure has a `returns table...`.   
This is fine, just making sure.

File: docs/kuneiform/procedures.md [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+    $age int := 100;
+    for $row in delete from users
+        where age = $age
+        returning * {
+        // can access $row.id, $row.name, and $row.age
+        // for all deleted users
+    }
+}
+```
+
```
Comment (KwilLuke) [inline docs/kuneiform/procedures.md:182]: If you add `SQL Loops` above, add `Procedure Loops` here.

```suggestion

#### Procedure Loops
```

File: docs/kuneiform/procedures.md [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+To return a table from a procedure, you can simply return any SQL query that returns data:
+
+```ts
+procedure get_users() public view returns table(id uuid, name text, age int) {
+    return select * from users;
+}
+```
+
+#### Return Next
```
Comment (KwilLuke) [inline docs/kuneiform/procedures.md:255]: Does `return next` return the following iteration in the loop? Because your example makes it seem so.

**Review by KwilLuke — CHANGES_REQUESTED** [review]
My feedback:

Items Missing:
- Foreign Procedure Documentation
- New data types
- New contextual variables (e.g. @txid)
- Built-in function documentation

What should change:
- Move `Contextual variables` to `Supported Features`.
- Minor typos, grammar, and clarity feedback (I proposed changes for all of these).

**Comment by jchappelow** [comment]
What complete kf schema would you suggest a good demo of procedures?  More than hello world, but not crazy complex.  One of the integ test schemas? @brennanjl

File: docs/kuneiform/sql-as-understood-by-kwil/functions.mdx [hunk 12]
```diff
@@ -0,0 +1,206 @@
+---
+sidebar_position: 2
+sidebar_label: "Supported Functions"
+id: supported-functions
+title: Supported SQL Functions
+description: Supported SQL functions in Kuneiform.
+slug: /kuneiform/functions
+---
+
```
Comment (KwilLuke) [inline docs/kuneiform/sql-as-understood-by-kwil/functions.mdx:9]: Add a note that functions can only be called in procedures?

File: docs/kuneiform/supported-features.mdx [hunk 13]
```diff
@@ -42,7 +42,7 @@ Kwil supports the following index types defined at the `table` level:
 
 ## Access Tags
 
-Access tags are used to control access to actions.  They are defined in the `action` signature.  All actions must specify `public` or `private`, and can additionally
+Access tags are used to control access to actions.  They are defined in the `action` and `procedure` signature.  All actions/procedures must specify `public` or `private`, and can additionally
 specify other tags.  Below is a list of supported tags:
 
 | Tag | Description |
```
[the remark this answers is not in the document]
Comment (brennanjl) [reply docs/kuneiform/supported-features.mdx:48]: Yeah it can

File: docs/kuneiform/procedures.md (hunk 1 above)
Comment (brennanjl) [reply docs/kuneiform/procedures.md:53]: I sort've understand what you're saying, but I am not sure how I would document this, without documenting every possible typecast

File: docs/kuneiform/procedures.md (hunk 2 above)
Comment (brennanjl) [reply docs/kuneiform/procedures.md:72]: Will document

File: docs/kuneiform/procedures.md (hunk 3 above)
Comment (brennanjl) [reply docs/kuneiform/procedures.md:83]: Yep, just added a page for this

File: docs/kuneiform/procedures.md (hunk 4 above)
Comment (brennanjl) [reply docs/kuneiform/procedures.md:104]: We currently do not support that

File: docs/kuneiform/procedures.md (hunk 6 above)
Comment (brennanjl) [reply docs/kuneiform/procedures.md:199]: Nested breaks do work like any other language, and yes we do support nested loops. We do not support continue, however it would be easy to add.

File: docs/kuneiform/procedures.md (hunk 9 above)
Comment (brennanjl) [reply docs/kuneiform/procedures.md:263]: Yeah, there is no `return collected`. A procedure that returns a table doesn't technically have to return anything (it would just be an empty table)

File: docs/kuneiform/procedures.md [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+sidebar_label: "Procedures"
+id: kuneiform-procedures
+title: Kuneiform Procedures
+description: An outline of Kuneiform procedures.
+slug: /kuneiform/procedures
+---
+
+Kuneiform supports a basic procedural language for building logic into schemas. Procedures can be used for building access control logic, performing arithmetic, and managing control flow.
+
+## Procedure Blocks
```
[the remark this answers is not in the document]
Comment (brennanjl) [reply docs/kuneiform/procedures.md:12]: I'll make it `Declaring Procedures`

File: docs/kuneiform/procedures.md [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+
+procedure iterate_range_dynamic($start int, $end int) public {
+    // can also iterate over dynamic ranges
+    for $i in $start..$end {
+        // ...
+    }
+}
+```
+
+#### Procedure Loops
```
[the remark this answers is not in the document]
Comment (brennanjl) [reply docs/kuneiform/procedures.md:160]: Ah whoops. We don't have procedure calls anymore, it's just `select * from procedure()`

File: docs/kuneiform/sql-as-understood-by-kwil/functions.mdx (hunk 12 above)
Comment (brennanjl) [reply docs/kuneiform/sql-as-understood-by-kwil/functions.mdx:9]: They can be called in select statements. Will add some docs to actions

**Comment by brennanjl** [comment]
Ok, I added new docs and addressed all comments

File: docs/kuneiform/introduction.mdx [hunk 16]
```diff
@@ -31,18 +27,61 @@ table users {
...
+   table users {
+       id int primary,
+       names text[] notnull
+   }
+```
+
+## Contextual Variables
+
+Kuneiform supports "contextual variables", which allow you to access information about the current transaction.
+Contextual variables can be accessed anywhere regular variables can be accessed, and are prefixed with an `@` symbol.
```
Comment (KwilLuke) [inline docs/kuneiform/introduction.mdx:69]: Should we specify **where** regular variables can be accessed?

File: docs/kuneiform/ddl.md [hunk 17]
```diff
@@ -73,22 +88,22 @@ table posts {
...
+- ```set null``` The column is set to null.
+- ```set default``` The column is set to the column's default value (if specified in the table declaration).
 - ```cascade``` The change in the reference column cascades to the specified column.  In the case of a deletion, it will delete the entire record.
 
 ## Indexing in Tables
 
-The Kuneiform Language not only allows column declarations, but also the creation of indexes on one or more columns within your table. Kuneiform supports standard SQLite B-tree indexes. For more details on SQLite indexes, please refer to [this tutorial](<https://www.sqlitetutorial.net/sqlite-index/>).
+The Kuneiform Language not only allows column declarations, but also the creation of indexes on one or more columns within your table. Kuneiform supports standard B-tree indexes.
 
 ## Index Declaration
```
Comment (KwilLuke) [inline docs/kuneiform/ddl.md:108]: I think "Index Declaration" should be contained within "Indexing in Tables" (as opposed to at the same level).

```suggestion
### Index Declaration
```

File: docs/kuneiform/ddl.md [hunk 18]
```diff
@@ -123,4 +138,12 @@ table users {
...
 
+Kwil supports the following index types defined at the `table` level:
+
+| Index Declaration | Description |
+| ----------------- | ----------- |
+| index(`<column-name-1>`, `<coloumn-name-2>`, ...) | Standard binary tree index |
+| unique(`<column-name-1>`, `<coloumn-name-2>`, ...) | Creates a binary tree index while enforcing the unique attribute on the specified columns |
+| primary(`<column-name-1>`, `<coloumn-name-2>`, ...) | Creates a binary tree index while enforcing the primary key values of the specified columns |
+
 This concludes the guide to table, column, and index declarations in the Kuneiform Language. With this knowledge, you're equipped to define and structure your data tables effectively.
```
Comment (KwilLuke) [inline docs/kuneiform/ddl.md:149]: IMO, we can remove this.

File: docs/kuneiform/procedures.mdx [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+slug: /kuneiform/procedures
+---
+
+Kuneiform supports a basic procedural language for building logic into schemas. Procedures can be used for building access
+control logic, performing arithmetic, and managing control flow.
+
+## Declaring Procedures
+
+Procedures are declared using the `procedure` keyword. They can be given a list of parameters and their corresponding types,
+return types, and access modifiers.
```
Comment (KwilLuke) [inline docs/kuneiform/procedures.mdx:16]: Add a note that if a procedure returns data, the return must be declared in the signature?

File: docs/kuneiform/procedures.mdx [hunk 20]
```diff
@@ -0,0 +1,336 @@
...
+// and therefore can only be called with a blockchain transaction.
+procedure create_user($name text, $age int) public {
+    // procedure implementation
+}
+```
+
+## Access Modifiers
+
+Procedures can be given access modifiers to specify how/when they can be called. All procedures must be specified
+as `public` or `private`: `public` procedures can be called externally, while `private` procedures can only be called
```
Comment (KwilLuke) [inline docs/kuneiform/procedures.mdx:57]: I think it is easier to read if the access modifiers are presented in a table. The way this reads, it is hard to follow all the supported tags.

File: docs/kuneiform/procedures.mdx [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+    // can also iterate over dynamic ranges
+    for $i in $start..$end {
+        // ...
+    }
+}
+```
+
+#### SQL Loops
+
+You can iterate over any SQL statement as well. The following examples reference an example table `users`, with columns `id`, `name`, and `age`.
```
Comment (KwilLuke) [inline docs/kuneiform/procedures.mdx:232]: This seems more precise:

```suggestion
You can also iterate over each record returned from a SQL query. The following examples reference an example table `users`, with columns `id`, `name`, and `age`.
```

File: docs/kuneiform/procedures.mdx [hunk 22]
```diff
@@ -0,0 +1,336 @@
...
+```ts
+procedure get_users() public {
+    $count int := 0;
+    for $row in SELECT * FROM users {
+        // do something
+
+        // break if too many records have been read
+        $count := $count + 1;
+        if $count >= 50 {
+            break
```
Comment (KwilLuke) [inline docs/kuneiform/procedures.mdx:269]: Does this `break` need a semicolon? It feels weird that every other line in this language ends with a semicolon, but `break` does not?

File: docs/kuneiform/foreign_procedures.mdx [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+foreign procedure get_user($id uuid) returns (text, int)
+
+// get_users is capable of calling a procedure in any other database on the network
+// that takes no arguments and returns a table with columns username and age
+foreign procedure get_users() returns table(username text, age int)
+```
+
+## Calling Foreign Procedures
+
+Foreign procedures can be called like any other function/procedure. The only difference is that the target
```
Comment (KwilLuke) [inline docs/kuneiform/foreign_procedures.mdx:28]:
```suggestion
Foreign procedures are called like any other function/procedure. The only difference is that the target
```

File: docs/kuneiform/foreign_procedures.mdx [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+```typescript
+foreign procedure get_user($id uuid) returns (text, int)
+
+procedure call_get_user($id uuid) public  {
+    $username, $age = get_user['x_target_dbid', 'target_procedure']($id);
+    // $username is a text and $age is an int
+}
+```
+
+Since the target procedure name is passed as a string, it is only necessary that the arguments and returns
```
Comment (KwilLuke) [inline docs/kuneiform/foreign_procedures.mdx:40]: wdym by "only"?

```suggestion
Since the target procedure name is passed as a string, it is necessary that the arguments and returns
```

File: docs/kuneiform/sql-as-understood-by-kwil/functions.mdx [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+:::info
+Most of these functions are direct copies of PostgreSQL functions, and further documentation can be found in the
+[PostgreSQL Documentation](https://www.postgresql.org/docs/16/functions.html). Functions that are not direct translations
```
Comment (KwilLuke) [inline docs/kuneiform/sql-as-understood-by-kwil/functions.mdx:12]:
```suggestion
[PostgreSQL Documentation](https://www.postgresql.org/docs/16/functions.html). Functions that are not direct translations from Postgres
```

File: docs/concepts.mdx [hunk 26]
```diff
@@ -23,29 +23,31 @@ Accounts are identified by public keys. Natively, Kwil supports Secp256k1 keys (
...
 
 [Kuneiform](/docs/category/kuneiform) is a simple syntax language for defining a database on Kwil. You can think of Kuneiform as a smart contract language (e.g. [Solidity](https://soliditylang.org/)), but for relational databases.
 
 In a Kuneiform file, you can define:
 
-- Relational Tables
-- Actions - CRUD Operations that can be executed on the database.
+- Tables - The structure of the database.
+- Procedures - SQL Queries and business logic that can be executed on the database.
+- Actions - CRUD Operations that can be executed on the database, and call extensions.
```
Comment (KwilLuke) [inline docs/concepts.mdx:36]: Hyperlink to extensions
```suggestion
- Actions - CRUD Operations that can be executed on the database, and call [extensions](#extensions).
```

File: docs/kuneiform/introduction.mdx (hunk 16 above)
Comment (brennanjl) [reply docs/kuneiform/introduction.mdx:69]: I don't really know how... if they are curious about how to access variables, they should just look at the procedure language docs.

Comment (brennanjl) [reply docs/kuneiform/introduction.mdx:69]: Like I am not sure how I would convey where variables can be used without just showing them how to use the language

File: docs/kuneiform/procedures.mdx (hunk 19 above)
Comment (brennanjl) [reply docs/kuneiform/procedures.mdx:16]: I feel like that is sort've implied...

I'll add a note about strong typing

File: docs/kuneiform/procedures.mdx (hunk 20 above)
Comment (brennanjl) [reply docs/kuneiform/procedures.mdx:57]: I sort've disagree, because a table conveys that they are all options. For example, in the table, since `view` and `owner` are there, it makes me think that I can use those instead of public. We even saw fractal trip up on this. I think this more clearly conveys that `owner` and `view` are not replacements for public and private

**Review by KwilLuke — APPROVED** [review]
